### PR TITLE
Add more endpoints to API

### PR DESCRIPTION
**Pull request by petebachant**
I also added a Makefile and target to hot reload for development.

### CODE REVIEW

File: api/src/api/users.rs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            return HttpResponse::InternalServerError().body(format!(
+                "failed to query database for existing user. error: {}",
+                e
+            ));
+        }
+    }
+
+    // create a new user document
+    // first, hash password
+    // TODO: permissions?
```
Comment (petebachant) [inline api/src/api/users.rs:54]: @Theodlz is there already a permissions model for this application, or is this not important at this time?

File: api/src/api/users.rs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+                                .map(|val| val.to_string())
+                                .unwrap_or_default(),
+                            email: doc
+                                .get_str("email")
+                                .map(|val| val.to_string())
+                                .unwrap_or_default(),
+                            id: doc
+                                .get_object_id("_id")
+                                .map(|id| id.to_string())
+                                .unwrap_or_default(),
```
Comment (jesaerys) [inline api/src/api/users.rs:106]: I wonder if `unwrap_or_default` is the right choice. How severe/critical/interesting is it to encounter an Err from these `.get_*` methods?

File: api/src/main.rs [hunk 3]
```diff
@@ -1,9 +1,17 @@
...
 
-use actix_web::{App, HttpServer, web};
+use actix_web::{App, HttpResponse, HttpServer, get, web};
 use mongodb::{Client, Database};
 
+#[get("/")]
+pub async fn get_health() -> HttpResponse {
+    HttpResponse::Ok().json(serde_json::json!({
+        "status": "success",
+        "message": "Greetings from BOOM!"
```
Comment (jesaerys) [inline api/src/main.rs:11]: BOOM in uppercase monospace reminds me of the DOOM logo... potential branding opportunity?

**Review by jesaerys — COMMENTED** [review]
Just some minor comments. I'm quite out of practice with actix-web, but it seems pretty straightforward. I'll defer to @Theodlz for approval.

File: api/src/main.rs (hunk 3 above)
Comment (petebachant) [reply api/src/main.rs:11]: Adapted from https://emojicombos.com/doom-ascii-art:

🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧🟧
🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛⬛🟧
🟧🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧🟧
⬜🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬛⬛⬛⬛🟧⬛⬛⬛⬛🟧⬜
⬜🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬜
⬜🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬜
⬜🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬜
⬜🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬜
⬜🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬜
⬜🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬜
⬜🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬜
⬜🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬜
⬜🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛⬛⬛⬛⬛⬛⬛🟧⬜
⬜🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧⬛🟧⬛⬛⬛🟧⬜
⬜🟧⬛⬛⬛🟧🟧⬛⬛⬛⬛🟧⬛⬛⬛⬛🟧🟧⬛⬛⬛🟧⬛⬛⬛🟧🟧⬛⬛⬛⬛🟧⬛⬛⬛🟧🟧🟧⬛⬛⬛🟧⬜
⬜🟧⬛⬛⬛🟧⬛⬛⬛⬛🟧🟧🟧⬛⬛⬛⬛⬛⬛⬛⬛🟧⬛⬛⬛⬛⬛⬛⬛⬛🟧🟧🟧⬛⬛🟧⬜🟧⬛⬛⬛🟧⬜
⬜🟧⬛⬛⬛⬛⬛⬛⬛🟧🟧⬜🟧🟧⬛⬛⬛⬛⬛⬛🟧🟧🟧⬛⬛⬛⬛⬛⬛🟧🟧⬜🟧⬛⬛🟧⬜🟧⬛⬛⬛🟧⬜
⬜🟧⬛⬛⬛⬛⬛⬛🟧🟧⬜⬜⬜🟧🟧⬛⬛⬛⬛🟧🟧⬜🟧🟧⬛⬛⬛⬛🟧🟧⬜⬜🟧🟧⬛🟧⬜🟧⬛⬛⬛🟧⬜
⬜🟧⬛⬛⬛⬛⬛🟧🟧⬜⬜⬜⬜⬜🟧🟧⬛⬛🟧🟧⬜⬜⬜🟧🟧⬛⬛🟧🟧⬜⬜⬜⬜🟧🟧🟧⬜🟧⬛⬛⬛🟧⬜
⬜🟧⬛⬛⬛⬛🟧🟧⬜⬜⬜⬜⬜⬜⬜🟧🟧🟧🟧⬜⬜⬜⬜⬜🟧🟧🟧🟧⬜⬜⬜⬜⬜⬜⬜⬜⬜🟧🟧⬛⬛🟧⬜
⬜🟧⬛⬛⬛🟧🟧⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜🟧⬛⬛🟧⬜
⬜🟧⬛⬛🟧🟧⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜🟧🟧⬛🟧⬜
⬜🟧⬛🟧🟧⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜🟧⬛🟧⬜
⬜🟧🟧🟧⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜⬜🟧🟧🟧⬜

File: api/src/api/users.rs (hunk 2 above)
Comment (petebachant) [reply api/src/api/users.rs:104]: It would be a problem if there were ever a user in the database without a username, email, or id, so it would at least be good if the caller saw a descriptive error and could go clean things up.

Comment (jesaerys) [reply api/src/api/users.rs:104]: Ok so I guess the question is, does that warrant a 4xx or 5xx response to the caller or should it be logged as a warning for the operators to see? I don't know what the right answer is.

File: api/src/api/users.rs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+#[derive(serde::Serialize)]
+struct User {
+    username: String,
+    email: String,
+    id: String,
+}
+
+#[get("/users")]
+pub async fn get_users(db: web::Data<Database>) -> HttpResponse {
+    let user_collection: Collection<mongodb::bson::Document> = db.collection("users");
```
[the remark this answers is not in the document]
Comment (petebachant) [reply api/src/api/users.rs:82]: Didn't know that was possible, but I will give it a shot.

File: api/src/api/users.rs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                .body(format!("failed to insert user into database. error: {}", e));
+        }
+    }
+}
+
+#[derive(serde::Serialize)]
+struct User {
+    username: String,
+    email: String,
+    id: String,
```
[the remark this answers is not in the document]
Comment (petebachant) [reply api/src/api/users.rs:77]: True, though Kowalski allows usernames to be changed, so a client would likely want to use an ID for that, i.e., `PUT /users/{user_id}`. Adding the unique index makes sense nonetheless.

File: api/src/api/users.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Theodlz) [reply api/src/api/users.rs:82]: I suggest looking at the code from the zvar-api repo I shared with you (over slack, some time ago I think, I can share the link again if needed), that's done in a lot of places there. You basically just need to give serialize/deserialize to the struct then you can just call `Collection<User>` and it "just works"** (**some edge cases but overall its very good).

**Comment by petebachant** [comment]
Strangely my test to ensure we can't insert duplicate users is failing here (allowing us to insert duplicate users) but passing on my local machine.

**Comment by Theodlz** [comment]
> Strangely my test to ensure we can't insert duplicate users is failing here (allowing us to insert duplicate users) but passing on my local machine.

It's based on your custom ID field right? Is a unique index on that field being created as part of your test?

File: api/src/api/users.rs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                id: user_id,
+                username: body.username.clone(),
+                email: body.email.clone(),
+            })
+            .unwrap(),
+        ),
+        // Catch unique index constraint error
+        Err(e) if e.to_string().contains("E11000 duplicate key error") => HttpResponse::Conflict()
+            .body(format!(
+                "user with username '{}' already exists",
```
Comment (petebachant) [inline api/src/api/users.rs:54]: This could technically be incorrect if we had a UUID collision, but I don't think that's going to happen with the number of users we expect here.

**Comment by petebachant** [comment]
> I mean if you want to avoid that, you can overwrite the _id by your custom uuid instead of adding another id. That way no weird mongo ids that are a pain to deserialize, and built in uniqueness based on _id? It's pretty much the same end result honestly so I don't mind one or the other.

I think the extra ID field is fine. What worries me is the state of the indexes left behind and not managed by some sort of schema migration tool. Similar things can happen with RDBs too.

In any case, I'm happy with where this is now. Would like to merge before moving on to cleaning up some of the other endpoints.

File: api/src/main.rs (hunk 3 above)
Comment (Theodlz) [reply api/src/main.rs:10]: We need a PR to put this at the top of the README

**Review by Theodlz — APPROVED** [review]
LGTM!
